infrastructure/tari_util/src/chacha20.rs

```diff
@@ -0,0 +1,269 @@
+    fn test_init_state() {
+        let key: [u8; 32] = [
+            0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11,
+            0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f,
+        ];
+        let nonce: [u32; 3] = [150994944, 1241513984, 0];
+        let counter: u32 = 1;
+        let state = construct_state(&key, &nonce, counter);
+        let desired_state: [u32; 16] = [
+            1634760805, 857760878, 2036477234, 1797285236, 50462976, 117835012, 185207048, 252579084, 319951120,
+            387323156, 454695192, 522067228, 1, 150994944, 1241513984, 0,
+        ];
+        assert_eq!(state, desired_state);
+    }
+
+    #[test]
+    fn test_chacha20_block() {
+        let state: [u32; 16] = [
+            1634760805, 857760878, 2036477234, 1797285236, 50462976, 117835012, 185207048, 252579084, 319951120,
+            387323156, 454695192, 522067228, 1, 150994944, 1241513984, 0,
+        ];
+        let block_state = chacha20_block(&state);
+        let desired_block_state: [u32; 16] = [
+            3840405776, 358169553, 534581072, 3295748259, 3354710471, 57196595, 2594841092, 1315755203, 1180992210,
+            162176775, 98026004, 2718075865, 3516666549, 3108902622, 3900952779, 1312575650,
+        ];
+        assert_eq!(block_state, desired_block_state);
+    }
+
+    #[test]
+    fn test_chacha20_cipher_keystream() {
+        let key: [u8; 32] = [
+            0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11,
+            0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f,
+        ];
+        let nonce: [u32; 3] = [0, 1241513984, 0];
+        let block_count: usize = 1;
+        let desired_cipher_bytes: Vec<u8> = vec![
+            0x22, 0x4f, 0x51, 0xf3, 0x40, 0x1b, 0xd9, 0xe1, 0x2f, 0xde, 0x27, 0x6f, 0xb8, 0x63, 0x1d, 0xed, 0x8c, 0x13,
+            0x1f, 0x82, 0x3d, 0x2c, 0x06, 0xe2, 0x7e, 0x4f, 0xca, 0xec, 0x9e, 0xf3, 0xcf, 0x78, 0x8a, 0x3b, 0x0a, 0xa3,
+            0x72, 0x60, 0x0a, 0x92, 0xb5, 0x79, 0x74, 0xcd, 0xed, 0x2b, 0x93, 0x34, 0x79, 0x4c, 0xba, 0x40, 0xc6, 0x3e,
+            0x34, 0xcd, 0xea, 0x21, 0x2c, 0x4c, 0xf0, 0x7d, 0x41, 0xb7,
+        ];
+        let cipher_bytes = chacha20_cipher_keystream(&key, &nonce, block_count);
+        assert_eq!(cipher_bytes, desired_cipher_bytes);
+    }
+
+    #[test]
+    fn test_encode_and_decode() {
+        let key: [u8; 32] = [
+            0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11,
+            0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f,
+        ];
+        let nonce: [u32; 3] = [0, 1241513984, 0];
+        // Text: "Ladies and Gentlemen of the class of '99: If I could offer you only one tip for the future, sunscreen
+        // would be it."
+        let data_bytes: Vec<u8> = vec![
```

Comment:
Use `.as_bytes().to_vec()` as you've done elsewhere in this PR

Reply:
When I do the ".as_bytes().to_vec()" the set of data bytes are not exactly the same as the example byte set provided by the IRTF 8439 RFC, these small changes will result in a different set of output bytes. In this case I think having the exact same testing data is useful.

Reply:
Ah ok - maybe add a comment to say this comes from IRTF 8439 RFC

Reply:
Comment in description of file.